### PR TITLE
Snapshot candidate: ERD-27, ERD-69, ERD-70, ERD-71, ERD-76, ERD-77

### PR DESCRIPTION
**Pull request by Waguramu**
**Goal**: prepare a snapshot release candidate.

**Scope**:
[ERD-27](https://github.com/ndsev/erdblick/issues/27) Toggle for Cesium OSM Layer
[ERD-69](https://github.com/ndsev/erdblick/issues/69) Encode camera/layer state in URL
[ERD-70](https://github.com/ndsev/erdblick/issues/70) Responsive layout
[ERD-71](https://github.com/ndsev/erdblick/issues/71) Improve map layer selection panel
[ERD-76](https://github.com/ndsev/erdblick/issues/76) Basic support of multiple styles
[ERD-77](https://github.com/ndsev/erdblick/issues/77) Persistent storage of camera, layer and styles
* Erdblick components and services refactoring

**Deliverables**:
* Refactored app component.
* Map panel component.
* Style service.

**How to test**:
1. Build and run as usual.

OR

1. Test via mapviewer build as usual.

**Notes**:
* ensure that config/config.json file and styles/* directory are present to test for multiple styles
* styles config expects the following structure:
```json
{
   "styles": [
       { "id": "style id", "url": "style url" }, ...
   ]
}
```

### CODE REVIEW

**Comment by MisterGC** [comment]
I started to review this combined PR - some initial remarks based on the screenshot below 
- Tested with Firefox on Ubuntu 22.04 using a 4k Monitor
- [ ] Fix the text position and size in the search field
- [ ] No focus indicator needed for the Show/Hide and focus buttons
- [ ] As you can see there is no initial style loaded by default, is this intended?
- [ ] I suggest to remove the headline ('Maps Layers')
- [ ] What about renaming 'Maps' to 'Maps & Layers'?
- [ ] The OSM ImageryProvider is also a layer, therefore I'd put it inside Maps & Layers too - later this could be a drop down that allows to select any of the supported (or a custom) ImageryProvider
image

**Comment by MisterGC** [comment]
- [ ] Replace this section of the README with a description about how it works now and how I can preconfigure multiple styles:

image

**Comment by MisterGC** [comment]
Although we can keep it for the upcoming snapshot - I'd say we definitely should work on the general layout - have a look how far away search'n'settings and inspection from each other (Firefox, 4k Monitor, 100% Zoom)
 
image

**Comment by MisterGC** [comment]
Some more suggestions for improvements: 
- [ ] Instead of the visibility icon and a on/off slider we should just use one icon which clearly indicate activated/deactivated for both OSM and mapget provided map feature layer
- [ ] Fix the the too dense layout, give it a little more air - at least the scrollbar overlap and the tile level selection clipping and the non- clearly visible separation between OSM and the other layer are clear issues

To keep in mind for the future: 
- we'll add support for transparancy for any layer, think about how this could be easily covered without the need to totally break the GUI
- [ ] 
- [ ] 

image

**Comment by MisterGC** [comment]
- [ ] After re-using the URL in the browser I need to change MapViewer viewport before I see any content
- [ ] Deactivation of Layers has no effect (toggling the 'eye' button)

**Comment by Waguramu** [comment]
- [ ]  Make sure that the styles get reloaded from file on reload.
- [ ]  Deactivate the default style when there are more styles available.
- [ ]  Add a possibility to merge external configs into the existing config when run in Docker?

**Comment by MisterGC** [comment]
- [ ] I can now expand/collapse nodes in the inspection panel by clicking on the row, but this prevents me from copying the values of cells, if I mark text, the row expands/collapses and the text gets unmarked - user should be able to do both copy and collapse/expand by clicking on the row
- [ ] URL encoded camera viewport should use WGS84 coords (not global 3d coords!?)

**Comment by Waguramu** [comment]
- [] Fix style toggled state on reload (always active but not applied yet?)

**Comment by Waguramu** [comment]
WON'T FIX:
- Allow jumpto target selection of targets via keyboard (substantial effort -> create an issue)
- consider filter the list instead of greying out (will lead to confusion and more Angular lifecycle triggers)

**Review by josephbirkner — CHANGES_REQUESTED** [review]
Took a good long look, found a few things to take care of.

**Comment by MisterGC** [comment]
Feedback regarding style configuration:

**Must-Haves**
- [ ] Default style cannot be refreshed - make it refreshable
- [ ] The erdblick docs talk about automatically picked up yaml-files and the example config does not use the `yaml` extension for the url part, but it is needed
     e.g. `my-style.yaml` needs an entry with the extension: `"url": "my-style.yaml"` - change the docs that the file + ext has to be provided
- [ ] Having a broken style, leads to a white screen the corresponding style id is shown in the UI, the console gives no explicit information that what went wrong
     in the scope of the parsing - at least show the yaml parsing error in the console, later we will intro a proper error reporting
- [ ] Providing a non-valid style URL still shows the style with the id, but the MapViewer is broken, console inspection is needed to understand there is an error
     - considering the docs are fixed, I'd say the style config should just not show up if the URL is invalid or even better show the style with `<style-id> (ERR: see browser console)`

**Nice-To-Have**
- [ ] If the `config.yaml` is broken the MapViewer falls back to using the default style only, which is good, but the user doesn't know if he has entered a wrong
     style config
- [ ] Disabling and Enabling style seems to make the style the top-level one, so depending on the order of enabling/disabling different styles may be on top
- [ ] Why is importing a style a two-click action, `+Import`, `Choose File`, `Confirm Import` - imo it would be no problem to import the style directly after selecting it

File: erdblick_app/app/erdblick.model.ts [hunk 1]
```diff
@@ -147,35 +163,191 @@ export class ErdblickModel {
...
         // Continue visualizing tiles with a delay.
         const delay = this.tileVisualizationQueue.length ? 0 : 10;
         setTimeout((_: any) => this.processVisualizationTasks(), delay);
     }
 
-    reloadStyle() {
-        // Delete the old style if present.
-        if (this.style)
-            this.style.delete();
+    loadStyles() {
```
[the remark this answers is not in the document]
Comment (Waguramu) [reply erdblick_app/app/erdblick.model.ts:178]: Needs to be addressed separately

File: erdblick_app/app/erdblick.model.ts [hunk 2]
```diff
@@ -147,35 +163,191 @@ export class ErdblickModel {
...
+                            this.renderTileLayer(tileLayer, this.styles!.get(styleId), styleId);
+                        }
+                    }
+                });
+            }
+        }
+        console.log(`Reloaded style: ${styleId}.`);
+    }
+
+    cycleImportedStyle(styleId: string, remove= false) {
```
[the remark this answers is not in the document]
Comment (Waguramu) [reply erdblick_app/app/erdblick.model.ts:263]: Needs to be addressed separately

File: erdblick_app/app/erdblick.model.ts [hunk 3]
```diff
@@ -54,13 +57,19 @@ export class ErdblickModel {
...
 
-    constructor(coreLibrary: any) {
+    private textEncoder: TextEncoder = new TextEncoder();
+
+    constructor(coreLibrary: any,
+                public styleService: StyleService,
+                public parametersService: ParametersService) {
         this.coreLib = coreLibrary;
-        this.style = null;
+        this.styles = null;
```
[the remark this answers is not in the document]
Comment (Waguramu) [reply erdblick_app/app/erdblick.model.ts:74]: Needs to be addressed separately

File: erdblick_app/app/erdblick.model.ts [hunk 4]
```diff
@@ -121,12 +131,18 @@ export class ErdblickModel {
...
 
             let [message, messageType] = this.tileStreamParsingQueue.shift();
+            console.log("PROCESS TILE STREAM: ", message, messageType)
             if (messageType === Fetch.CHUNK_TYPE_FIELDS) {
                 uint8ArrayToWasm(this.coreLib, (wasmBuffer: any) => {
                     this.tileParser.readFieldDictUpdate(wasmBuffer);
                 }, message);
             } else if (messageType === Fetch.CHUNK_TYPE_FEATURES) {
-                this.addTileFeatureLayer(message.slice(Fetch.CHUNK_HEADER_SIZE), null, null);
+                this.styles?.forEach((style, styleId) => {
```
[the remark this answers is not in the document]
Comment (Waguramu) [reply erdblick_app/app/erdblick.model.ts]: Needs to be addressed separately

File: erdblick_app/app/map.service.ts [hunk 5]
```diff
@@ -1,18 +1,61 @@
...
+//
+// export interface ErdblickLayer {
+//     name: string;
+//     coverage: BigInt;
+//     level: number;
+//     visible: boolean;
+// }
 
 @Injectable({providedIn: 'root'})
 export class MapService {
```
[the remark this answers is not in the document]
Comment (Waguramu) [reply erdblick_app/app/map.service.ts:31]: Needs to be addressed separately

File: angular.json [hunk 6]
```diff
@@ -56,8 +56,8 @@
                                 "output": "/bundle/images"
                             },
                             {
-                                "glob": "config/config.json",
-                                "input": ".",
+                                "glob": "**/*.json",
```
[the remark this answers is not in the document]
Comment (Waguramu) [reply angular.json:59]: They are?

**Review by josephbirkner — APPROVED** [review]
Ok, I think we can merge this now :)

Here is my list of refactoring needs:

* Find a way to ensure that coreLib loading happens before anything else, and that coreLib access is hassle-free from everywhere, probably using an Angular Resolver (https://chat.openai.com/share/e/981edbfc-01c3-4fa8-b121-f519bf7b18ba).
* Migrate style parsing from `ErdblickModel` to `StyleService`.
* Merge ErdblickStyle/ErdblickStyleData/StyleService.styles/StyleService.importedStyles.
* Merge Model and MapService.
* Migrate URL parameter handling from AppComponent to ParameterService.
* Turn ErdblickView into a real component or service?
* visualization is currently a `.component.ts`, but that is misleading.

**Pull request merged**